lib/circuit/CircuitState/CircuitState.ts

```diff
@@ -0,0 +1,74 @@
+import { Policy } from '../Policy/Policy';
+
+export class CircuitState {
+
+  private state: 'open' | 'close' | 'half-open' = 'open';
+  private timers: number[] = [];
+
+  constructor(
+    private readonly timeout: number,
+    private readonly interval: number,
+    private readonly errorsFilter: (error: Error) => boolean,
+    private readonly policy: Policy,
+  ) { }
+
+  public allowExecution(): boolean {
+    return this.state !== 'close';
+  }
+
+  public register(error?: Error): this {
+    const isError = error && this.errorsFilter(error);
+    const type = isError ? 'error' : 'success';
+
+    if (this.state === 'half-open') {
+      if (isError) {
+        this.close();
+      } else {
+        this.open();
+      }
+    }
+
+    this.policy.addExecution(type);
+    this.removeExecution(type);
+
+    if (!this.policy.allowExecution()) {
+      this.close();
+    }
+
+    return this;
+  }
+
+  private removeExecution(type: 'success' | 'error'): void {
```

Comment:
`removeExecution` name is misleading in my opinion. It does not remove it, rather it registers it for later removal. A name like `registerCall` would be more appropriate (or maybe smth else, but not `remove`).